### PR TITLE
[new Gameplay] AttackerTactic + Skills

### LINKED ISSUE #3072: [Gameplay] Create Skill interface
### Description of the task

Create Skill interface. Incorporate dummies for DribbleSkill, ShootSkill, WingerSkill and PassSkill.

(more info in the [RFC](https://www.notion.so/ubcthunderbots/Gameplay-Architecture-RFC-6db55ea5d63e422e9a69c4ce6457840d?pvs=4))

### Acceptance criteria

- [ ] Create Skill interface

### LINKED ISSUE #3073: [Gameplay] Implement DribbleSkill
### Description of the task

- Refactor DribbleFSM into DribbleSkill
- Replace DribbleFSM uses with DribbleSkill

### Acceptance criteria

- [ ] Refactor DribbleFSM into DribbleSkill
- [ ] Replace uses of DribbleFSM with DribbleSkill
- [ ] Migrate DribbleFSM tests into DribbleSkill

### Blocked By

#3072

### LINKED ISSUE #3074: [Gameplay] Implement ShootSkill
### Description of the task

Create ShootSkill to shoot at the net

### Acceptance criteria

- [ ] Implement ShootSkill
- [ ] Create simulated PyTests for ShootSkill

### LINKED ISSUE #3076: [Gameplay] Implement Pass skill
### Description of the task

Implement a PassSkill to pass the ball using a FSM.

### Acceptance criteria

- [ ] Create a PassSkill
- [ ] Write a simulated PyTest for the new PassSkill

### Blocked By

#3072

### PR DESCRIPTION
**Pull request by itsarune**
## Please fill out the following before requesting review on this PR

**This PR is not yet ready to be merged**. There are still some performance issues I am debugging and some segfaults. I am opening it right now, to unblock some future work.

### Description

This PR handles:
- AttackerTactic skill selection
- Implementation of PassSkill, ShootSkill and DribbleSkill
- Next skill selection using `calculateViability`: no scoring
- runs PassGenerator on its own thread

### Resolved Issues

resolves #3076 
resolves #3072 
resolves #3074 
resolves #3073 

### Length Justification and Key Files to Review

Please pay attention to the `software/ai/hl/stp/tactic/attacker/' folder

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Comment by williamckha** [comment]
~~I'm going to undraft this so that I can see the diff on github~~
nvrmind that didnt do anything

**Comment by williamckha** [comment]
Merged in changes from `new_gameplay_staging` and resolved conflicts

File: src/software/ai/hl/stp/strategy/strategy.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        std::unique_lock<std::mutex> lock(world_lock_);
+        world_available_cv_.wait(lock, [&] { return current_world_.has_value(); });
+    }
+
+    while (!end_analysis_)
+    {
+        std::scoped_lock lock(pass_generator_lock_, world_lock_);
+        latest_pass_eval_ = std::make_shared<PassEvaluation<EighteenZoneId>>(
+                pass_generator_->generatePassEvaluation(current_world_.value()));
+        pass_available_cv_.notify_one();
```
Comment (nimazareian) [inline src/software/ai/hl/stp/strategy/strategy.cpp]: We should update the locked regions to cover as little as possible. e.g. This seems to help performance
```suggestion
        std::optional<World> passing_world;
        {
            std::scoped_lock lock(world_lock_);
            passing_world = current_world_.value();
        }
        auto pass = std::make_shared<PassEvaluation<EighteenZoneId>>(
                pass_generator_->generatePassEvaluation(passing_world.value()));
        {
            std::scoped_lock lock(pass_generator_lock_);
            latest_pass_eval_ = pass;
        }
        pass_available_cv_.notify_one();
```

**Comment by itsarune** [comment]
will be merged via #3131

**Pull request closed without merging**